docs/specification/new-spec/consensus-reactor.md

```diff
@@ -0,0 +1,239 @@
+	Precommits               cmn.BitArray        // All precommits peer has for this round
+	LastCommitRound          int                 // Round of commit for last height. -1 if none.
+	LastCommit               BitArray            // All commit precommits of commit for last height.
+	CatchupCommitRound       int                 // Round that we have commit for. Not necessarily unique. -1 if none.
+	CatchupCommit            BitArray            // All commit precommits peer has for this height & CatchupCommitRound
+}
+```   
+
+## Receive method of Consensus reactor
+
+When a message is received from a peer p, it is treated differently based on the channel message belongs to.
+
+TODO: Add high level description of the logic of Receive method. 
+
+## Gossip Data Routine
+
+It is used to send the following messages to the peer: `BlockPartMessage`, `ProposalMessage` and 
+`ProposalPOLMessage` on the DataChannel. The gossip data routine is based on the local RoundState (denoted rs) 
+and the known PeerRoundState (denotes prs). The routine repeats forever the logic shown below:
+
+```
+1a) if rs.ProposalBlockPartsHeader == prs.ProposalBlockPartsHeader and the peer does not have all the proposal parts then
+        Part = pick a random proposal block part the peer does not have 
+        Send BlockPartMessage(Height, Round, Part) to the peer on the DataChannel 
+        if send returns true, record that the peer knows the corresponding block Part
+	    Continue  
+ 
+1b) if (0 < prs.Height) and (prs.Height < rs.Height) then
+        help peer catch up using gossipDataForCatchup function
+        Continue
+
+1c) if (rs.Height != prs.Height) or (rs.Round != prs.Round) then 
+        Sleep PeerGossipSleepDuration
+        Continue   
+
+//  at this point rs.Height == prs.Height and rs.Round == prs.Round
+1d) if (rs.Proposal != nil and !prs.Proposal) then 
+        Send ProposalMessage(rs.Proposal) to the peer
+        if send returns true, record that the peer knows Proposal
+	    if 0 <= rs.Proposal.POLRound then
+	    polRound = rs.Proposal.POLRound 
+        prevotesBitArray = rs.Votes.Prevotes(polRound).BitArray() 
+        Send ProposalPOLMessage(rs.Height, polRound, prevotesBitArray)
+        Continue  
+
+2)  Sleep PeerGossipSleepDuration 
+```
+
+## Gossip Votes Routine
+
+It is used to send the following message: `VoteMessage` on the VoteChannel.
+The gossip votes routine is based on  the local RoundState (denoted rs) 
+and the known PeerRoundState (denotes prs). The routine repeats forever the logic shown below:
+
+```
+1a) if rs.Height == prs.Height then
+        if prs.Step == RoundStepNewHeight then    
+            vote = random vote from rs.LastCommit the peer does not have  
+            Send VoteMessage(vote) to the peer 
+            if send returns true, continue
+  
+        if prs.Step <= RoundStepPrevote and prs.Round != -1 and prs.Round <= rs.Round then                 
```

Comment:
If we are lagging behind peer (prs.Round > rs.Round), we are not forwarding messages to him before we catch up. Let consider that p1 is in round 5, p2 in round 3 and p3 in round 5 and they are connected like this p1 <-> p2 <-> p3. p1 could send to p2 vote message with round 5 so p2 could catch up to round 5 and then p2 could be able to immediately forward message to p3 as it would be beneficial for p3 to receive this message in a timely fashion.

Reply:
That makes sense.

Reply:
But, we need to be aware of memory issues.  See the comment below.